### ISSUE TITLE
Is there a document which guides to convert the contextual bandit  model to format  like csv?

### ISSUE THREAD

**Issue by musram** [VowpalWabbit/vowpal_wabbit#3444]
### Description

> A brief description of the error, missing documentation or what you would like added
We can't host a contextual bandit model in our serving module as our serving module requires the model to be in CSV format. For a model like regression we use --inverted-hash and convert the model as a CSV format. Is there any document that can help contextual bandit model to convert into a format like CSV?

### Link to Documentation Page

> Where is the documentation in question?

**Comment by musram** [comment]
I  trained the contextual bandit as vw1 = pyvw.vw("-d data/cb_load.dat --cb_explore_adf -q UA  -P 1 --invert_hash mymodel.inverted") on https://github.com/VowpalWabbit/vowpal_wabbit/blob/master/test/train-sets/cb_load.dat  with the --inverted-hash and I got the mymodel.inverted as:

I could understand the user and action features. What does  18107:0.137426 . means in "User^time_of_day=afternoon*Action^article=politics:18107:0.137426" ?
I think 18107 is the hash value for "User^time_of_day=afternoon*Action^article=politics" and  0.137426 is weight. I don't know if this is correct? 

How can I get the probability corresponding to the user and action features from the weights?

Version 8.11.0
Id 
Min label:-1
Max label:0
bits:18
lda:0
0 ngram:
0 skip:
options: --cb_adf --cb_explore_adf --cb_type mtr --csoaa_ldf multiline --csoaa_rank --quadratic UA
Checksum: 2033437909
event_sum 113
action_sum 791
:0
User^time_of_day=afternoon*Action^article=politics:18107:0.137426
User^user=Tom:32581:-0.0636371
User^user=Tom*Action^article=politics:38087:-0.0636749
Action^article=politics:52568:-0.110663
User^time_of_day=morning*Action^article=music:58967:0.224528
User^user=Anna*Action^article=politics:62875:0.0165196
User^time_of_day=afternoon:65137:-0.0253498
Action^article=music:67569:-0.0505464
User^time_of_day=afternoon*Action^article=food:67793:0.121444
User^time_of_day=morning*Action^article=politics:77054:-0.192732
User^user=Anna*Action^article=music:81714:0.297336
Action^article=sports:86811:0.0540273
User^user=Tom*Action^article=music:89710:-0.101787
User^user=Anna*Action^article=sports:93144:0.0540273
Action^article=food:99122:0.121444
User^time_of_day=afternoon*Action^article=music:101394:-0.190187
User^user=Anna*Action^article=food:113649:0.0457554
Constant:116060:-0.0947514
User^time_of_day=afternoon*Action^article=sports:121080:0.0540273
User^user=Tom*Action^article=food:121517:0.109427
Action^article=camping:134640:0.0742112
User^user=Anna:141841:0.0966574
User^time_of_day=afternoon*Action^article=health:144204:0.0344906
User^user=Tom*Action^article=camping:152687:0.0742112
User^user=Anna*Action^article=health:163948:0.0344906
Action^article=health:178351:0.0971796
User^user=Tom*Action^article=health:188720:0.09161
User^time_of_day=morning*Action^article=health:219401:0.09161
User^time_of_day=morning:243586:0.0320462
User^time_of_day=morning*Action^article=camping:257110:0.0742112

**Comment by musram** [comment]
When I am using vw1 = pyvw.vw("-d data/cb_load.dat --cb_explore_adf -q UA  -P 2 --invert_hash mymodel.inverted -f saved_model1.model --save_resume") I am getting the output as   below.

why is the output different compared to vw1 = pyvw.vw("-d data/cb_load.dat --cb_explore_adf -q UA -P 1 --invert_hash mymodel.inverted") ?

I am not able to understand the the ouput for example  User^time_of_day=afternoon*Action^article=politics:18107:0.137426 28.0234 1 ?

Version 8.11.0
Id 
Min label:-1
Max label:0
bits:18
lda:0
0 ngram:
0 skip:
options: --cb_adf --cb_explore_adf --cb_type mtr --csoaa_ldf multiline --csoaa_rank --quadratic UA
Checksum: 2033437909
event_sum 113
action_sum 791
:1
initial_t 0
norm normalizer 376.654
t 113
sum_loss -36.5846
sum_loss_since_last_dump 0
dump_interval 114
min_label -1
max_label 0
weighted_labeled_examples 113
weighted_labels 0
weighted_unlabeled_examples 0
example_number 113
total_features 3503
total_weight 62.7757
sd::oec.weighted_labeled_examples 112
current_pass 1
User^time_of_day=afternoon*Action^article=politics:18107:0.137426 28.0234 1
User^user=Tom:32581:-0.0636371 33.811 1
User^user=Tom*Action^article=politics:38087:-0.0636749 16.0519 1
Action^article=politics:52568:-0.110663 35.6388 1
User^time_of_day=morning*Action^article=music:58967:0.224528 4.29297 1
User^user=Anna*Action^article=politics:62875:0.0165196 19.587 1
User^time_of_day=afternoon:65137:-0.0253498 38.1989 1
Action^article=music:67569:-0.0505464 11.8811 1
User^time_of_day=afternoon*Action^article=food:67793:0.121444 2.23671 1
User^time_of_day=morning*Action^article=politics:77054:-0.192732 7.6154 1
User^user=Anna*Action^article=music:81714:0.297336 0.47507 1
Action^article=sports:86811:0.0540273 0.279561 1
User^user=Tom*Action^article=music:89710:-0.101787 11.406 1
User^user=Anna*Action^article=sports:93144:0.0540273 0.279561 1
Action^article=food:99122:0.121444 2.23671 1
User^time_of_day=afternoon*Action^article=music:101394:-0.190187 7.5881 1
User^user=Anna*Action^article=food:113649:0.0457554 0.0777891 1
Constant:116060:-0.0947514 54.3015 1
User^time_of_day=afternoon*Action^article=sports:121080:0.0540273 0.279561 1
User^user=Tom*Action^article=food:121517:0.109427 2.15892 1
Action^article=camping:134640:0.0742112 1.54022 1
User^user=Anna:141841:0.0966574 20.4905 1
User^time_of_day=afternoon*Action^article=health:144204:0.0344906 0.07108 1
User^user=Tom*Action^article=camping:152687:0.0742112 1.54022 1
User^user=Anna*Action^article=health:163948:0.0344906 0.07108 1
Action^article=health:178351:0.0971796 2.72511 1
User^user=Tom*Action^article=health:188720:0.09161 2.65403 1
User^time_of_day=morning*Action^article=health:219401:0.09161 2.65403 1
User^time_of_day=morning:243586:0.0320462 16.1026 1
User^time_of_day=morning*Action^article=camping:257110:0.0742112 1.54022 1

**Comment by lokitoth** [comment]
> I could understand the user and action features. What does 18107:0.137426 . means in "User^time_of_day=afternoonAction^article=politics:18107:0.137426" ?
> I think 18107 is the hash value for "User^time_of_day=afternoonAction^article=politics" and 0.137426 is weight. I don't know if this is correct?

Yes, this is correct.

> How can I get the probability corresponding to the user and action features from the weights?

I am not sure what you mean by probability here. Do you mean how would you use the weights to generate a prediction for a specific instance of a CB problem?

> When I am using vw1 = pyvw.vw("-d data/cb_load.dat --cb_explore_adf -q UA -P 2 --invert_hash mymodel.inverted -f saved_model1.model --save_resume") I am getting the output as below.
>
> why is the output different compared to vw1 = pyvw.vw("-d data/cb_load.dat --cb_explore_adf -q UA -P 1 --invert_hash mymodel.inverted") ?
> 
> I am not able to understand the the ouput for example User^time_of_day=afternoon*Action^article=politics:18107:0.137426 28.0234 1 ?

With `--save_resume` added, the output will contain not just the model weights, but also the learning state information (which is used to resume learning "in place" if splitting between multiple runs of vw using `-f` and `-i`). On feature lines this adds additional values after `<feature_name>:<feature_index>:<weight_value>`. These values are specific to the base learner configuration. 

See [saving a model](Model-saving-and-loading#saving-a-model) and [invert_hash Output](invert_hash-Output) for more details about the output of `--invert_hash`.

**Comment by musram** [comment]
Thanks lokitoth for your reply.
Do you mean how would you use the weights to generate a prediction for a specific instance of a CB problem?
For example, if my strategy is epsilon-greedy, then let 
a = argmax<context_feature_specific_instance, model_weights_for_context_feature_and_action_a>
           a
 be the greedy action. and then use standard epsilon-greedy.  I think this is the way of prediction using epsilon-greedy. Pls, correct me if I am wrong.
For the other strategy like regcb, I have to read the paper and go through the code to understand the prediction method.

**Comment by lokitoth** [comment]
Depending on what your purpose is, your best bet might be to use VW to run the inference pass, since it already includes the code to do so without having to extract the values from the readable model, then reproduce the mathematics.

Since it looks like you are in python, and you already have the VW model loaded (or load from a generated loadable model file using `-f` and then `-i`), what you can do is call the predict() method on the `vw` object. 

There is a Python tutorial [here](https://vowpalwabbit.org/tutorials/contextual_bandits.html#python-tutorial) that shows how to construct an example and generate a prediction through python programmatically.

**Comment by musram** [comment]
Hi Jacob Alber lokitoth. 
Thanks for your reply.

My constraint is that I cant use the VW binary model in my serving system. My serving system supports only CSV format.  Hence I raised this issue of converting using --invert-hash and then using it as a CSV in my serving environment.

So I have to build the offline model and then convert it into CSV and then write code to do inference. 

Hope I am clear.

**Comment by lokitoth** [comment]
No worries, that makes sense. 

There are a few things you should keep in mind when doing this:

1. Make sure you have a good understanding of [how the hash-encoding works](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Feature-Hashing-and-Extraction).
2. You can get a list of the enabled reductions which are running in your stack via `vw.get_enabled_reductions()`. This should can help you look up the code of what VW does.
3. Generally speaking, you will be doing the following, given an example of the form { shared_features, [ actions_features ] }:
   1. For each action, generate the combined example by concatenating shared_features and action_features[i]
   2. Compute the [generated feature interactions](https://github.com/VowpalWabbit/vowpal_wabbit/wiki/Feature-interactions)
   3. Compute the score for the action
   4. Pick the best action - this will be the "exploit" action
   5. Run the appropriate exploration distribution generation code, given the exploit action.
   6. Sample out of the distribution (though if you want to have 1:1 reproduction of the code, make sure to follow the sampling logic in the [explore library](https://github.com/VowpalWabbit/vowpal_wabbit/blob/master/explore/explore.h))

**Comment by musram** [comment]
Hi Jacob Alber lokitoth.
Your checklist was great and helpful.

Regarding your question:
For a better sense of your constraints: Are you able to run the VW code?
Yes I am able to run the VW code

 Is it just the form that the model file has to take?
Yes.

Thanks a lot, Jacob Alber lokitoth.

**Comment by musram** [comment]
Hi Jacob Alber lokitoth, 

Is there any way to add  importance weight to off line training of the contextual bandits? Similar to linear regression where we specify   importance weight as 2  in the training example " 1 2 second_house | price:.18 sqft:.15 age:.35 1976".

This will help in reducing the training time of contextual bandits as the training data points are in billions.  But we get good reduction if we use importance weight as data points are repeated.

**Comment by olgavrou** [comment]
Hi @musram at the moment it isn't possible to set the importance weight on a cb label the same way it can be set on a simple label. I have created this feature request [here](https://github.com/VowpalWabbit/vowpal_wabbit/issues/3472). Closing this issue, but please feel free to reopen if you have more questions

**Comment by musram** [comment]
Thanks @olgavrou
